### ISSUE TITLE
Solver failed

### ISSUE THREAD

**Issue by Thuong-ironman** [Thuong-ironman/OCR_2023#5]
Hi @matteodv99tn , I now trained a neural network to predict the terminal cost at a arbitrary input state, and integrated it to MPC problem in here:

https://github.com/Thuong-ironman/OCR_2023/blob/9569ed1d3aa386bc6817cb0753a529d7ddf18f85/OCP_SinglePendulum.py#L123-L125

But when I run the MPC iteration, the solver did not find a solution, whereas using MPC without terminal cost, I am able to solve it.
Do you know why that happen and how to solve it?

**Comment by matteodv99tn** [comment]
Where are you evaluating the neural network exactly?

Your terminal cost in this case should be the value provided by the neural network you trained evaluated for the last state.

**Comment by Thuong-ironman** [comment]
Here I provided to the neural network the batch of input initial state x0 corresponding to the label cost J(x0)
https://github.com/Thuong-ironman/OCR_2023/blob/9569ed1d3aa386bc6817cb0753a529d7ddf18f85/train.py#L65-L66

and here:
https://github.com/Thuong-ironman/OCR_2023/blob/9569ed1d3aa386bc6817cb0753a529d7ddf18f85/train.py#L74C5-L77C52

When the training done, I load it here for evaluating new terminal cost:
https://github.com/Thuong-ironman/OCR_2023/blob/9569ed1d3aa386bc6817cb0753a529d7ddf18f85/OCP_SinglePendulum.py#L96C5-L108C17

**Comment by matteodv99tn** [comment]
I don't get how the following lines evaluate the NN model
https://github.com/Thuong-ironman/OCR_2023/blob/9569ed1d3aa386bc6817cb0753a529d7ddf18f85/OCP_SinglePendulum.py#L96-L108

I never properly used tensorflow, but this seems wrong. Shouldn't you call something like ``tf.predict(model, x)``?

**Comment by Thuong-ironman** [comment]
This is what Prof suggested due to the incompatibility of tensorflow data type with casadi. It's simply a substitution of using tf.predict(model,x).
![Screenshot from 2024-01-19 11-39-33](https://github.com/Thuong-ironman/OCR_2023/assets/19421386/b2b647ea-00b7-40d2-8b5a-02c72c61912b)

**Comment by matteodv99tn** [comment]
Why didn't you just copy-paste that code snippet? 
Moreover, did you used only relu functions in your NN while training?
In caso I don't think I can help you that much, these are things I never tried
